CORE-376 kafka/quota_mgr: fix client quota throttling to be node-wide

pgellert opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Group throttling limits do not behave as customers expect.  Customers expect that group throttling limits take affect across the entire node, that is if they set a throttle for a group to be 5MBps, all clients that match that group should be limited, in total to 5MBps on that node.  However, group throttling occurs per shard.  If two clients end up on one shard, and another client ends up on another shard, then those two shards will be throttled to 5MBps in aggregate and the third client will be throttled to 5MBps.

To fix this, this PR reuses the same primitives that we used to make ingress/egress throttling node-wide (that is `ss::internal::shared_token_bucket` and `sharded_ptr`) to make client quotas node-wide as well.

Fixes https://redpandadata.atlassian.net/browse/CORE-376

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.1.x
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -285,15 +351,35 @@ ss::future<throttle_delay> quota_manager::throttle_fetch_tp(
+          _gate,
+          [full_window,
+           _client_quotas = std::ref(_client_quotas)]() -> ss::future<void> {
+              return ss::smp::submit_to(
+                _client_quotas.get().shard_id(),
+                [full_window, _client_quotas = std::ref(_client_quotas)]()
+                  -> ss::future<void> {
```
I don't think it makes sense for every shard to dispatch a gc to the single shard. The tiimer can be set just on the worker shard,

pgellert replied:
Changing this to only fire from the owner shard

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -36,55 +55,114 @@ quota_manager::quota_manager()
+           now](const client_quotas_map_t& old_map) -> client_quotas_map_t {
+            auto new_map = client_quotas_map_t{old_map};
+
+            // Set the burst limit to the full window length to be similar to
+            // allow a similar level of bursting as Kafka's window-based
+            // algorithm
+            auto burst_limit_multiplier = default_num_windows
+                                          * default_window_width
+                                          / std::chrono::seconds(1);
```
I don't think this bursting is necessary as the bucket is effectively limitless and tokens can always be taken.

pgellert replied:
Tokens can be taken, but if the balance goes negative we start applying a delay -- the bursting prevents that and lets clients burst without delay.

I'm in two minds about whether that is useful or not, but that is the current behaviour and some tests start failing if we don't allow bursting because they depend on the bursting behaviour. But the tests only implicitly depend on this behaviour, specifically where they are producing 10 messages and assume that that's below the limit, whereas the request size of that is just above the limit, and only bursting allows those produce requests not to be throttled. So I think we could just reduce that 10 to ~8 if we don't want to allow bursting. Eg. here: https://github.com/redpanda-data/redpanda/blob/e89ed704260addd2f5398de874b2d9e134a4a9e7/tests/rptest/tests/cluster_quota_test.py#L192-L221

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -36,55 +55,114 @@ quota_manager::quota_manager()
 
 ss::future<> quota_manager::start() {
+    co_await seastar::smp::submit_to(
+      _client_quotas.shard_id(),
+      [_client_quotas = std::ref(_client_quotas)]() -> ss::future<> {
+          co_await _client_quotas.get().reset(client_quotas_map_t{});
+          co_return;
+      });
+
     _gc_timer.arm_periodic(_gc_freq);
-    return ss::make_ready_future<>();
+    co_return co_await ss::make_ready_future<>();
```
`start` is called on every shard, so I think it's best for the other shards to guard the whole contents of this function by the worker shard.

BenPope commented on tests/rptest/tests/cluster_quota_test.py:
```diff
@@ -176,7 +176,7 @@ def produce(self, producer, amount):
             f.get(1)
 
-    def fetch(self, consumer, messages_amount, timeout_sec=300):
+    def fetch(self, consumer, messages_amount, timeout_sec=15):
```
Why does this value change?

pgellert replied:
I changed it while testing to timeout faster on failure. Locally 15s was plenty, but I'll change it back to 300 just to avoid any CI flakiness.

BenPope commented on tests/rptest/tests/cluster_quota_test.py:
```diff
@@ -405,11 +405,12 @@ def test_client_response_throttle_mechanism_applies_to_next_request(self):
             client_id="consumer",
             consumer_timeout_ms=1000,
-            max_partition_fetch_bytes=self.target_default_quota_byte_rate * 10,
+            max_partition_fetch_bytes=self.break_default_quota_message_amount *
+            self.message_size,
```
Can you help me understand this change?

pgellert replied:
Added more context into the commit message now: https://github.com/redpanda-data/redpanda/pull/18098/commits/096f273b1ea3bf6c689eaaccbd1655f99d19400c

Originally this change was required because it also increased the data fetched from 10x to 11x the quota rate. If we allow bursting and have a 10x burst rate configured we need this because the token bucket gets refilled between the 2 fetches so we need to have the token bucket go enough into the negative to not replenish by the next fetch. This is now not required if we don’t allow bursting, but I think the change is worth keeping for the other 2 points shown in the commit message.

BenPope commented on tests/rptest/tests/cluster_quota_test.py:
```diff
@@ -405,11 +405,12 @@ def test_client_response_throttle_mechanism_applies_to_next_request(self):
             enable_auto_commit=False)
 
-        self.produce(producer, self.break_default_quota_message_amount)
+        self.produce(producer, self.break_default_quota_message_amount * 2)
```
I'm not following why this change is being made.

BenPope reviewed (commented):
Where are configurations for groups updated? Does that need to be limited to the worker shard?

nvartolomei commented on src/v/ssx/include/ssx/sharded_ptr.h:
```diff
@@ -83,12 +84,35 @@ class sharded_ptr {
+    requires std::invocable<F, const T&>
+             && std::same_as<T, std::invoke_result_t<F, const T&>>
+    ss::future<> update(F update_func) {
+        assert_shard();
```
nit: assert_shard message is a bit confusing `reset must be called on home shard: ` maybe drop the word reset from there

pgellert commented on src/v/kafka/server/atomic_token_bucket.h:
```diff
@@ -0,0 +1,78 @@
+    /// If there are less than `v` tokens in the bucket, this call will still
+    /// succeed, but subsequent calls to calculate_delay() will return a
+    /// positive delay.
+    void record(uint64_t v) { _bucket.grab(v); }
```
I was thinking about folding replenish into one of the methods, but I was thinking about folding it into `calculate_delay` rather than `record/grab`. I saw that `snc_quota_manager` calls `replenish` right after grab, but to me it would make more sense to call replenish right before checking for the delay to ensure the delay is up-to-date. What do you think?

pgellert commented:
> Where are configurations for groups watched and updated? Does that need to be limited to the worker shard? Does it need to copy the map and deal with the additions and removals?

Good question. It looks like we don't watch for config changes but only apply them when we first see a new client id (by always reading the latest config values). That's problematic for long-living client ids. I think we should fix that and actively watch for config changes, but I also think that it should be a separate PR. I'll make a ticket for it if it sounds good to you too.

BenPope commented on src/v/kafka/server/atomic_token_bucket.h:
```diff
@@ -0,0 +1,78 @@
+// atomic_token_bucket -- configures whether the bucket can grow indefinitely
+// (capped_release::no) or if the bucket is capped to a maximal burst rate set
+// by the limit constructor argument (capped_release::yes)
+template<ss::internal::capped_release capped_release>
```
> `snc_quota_manager` uses `capped_release::no` at the moment. However I think that might be a mistake and perhaps we should be using `capped_release::yes` there as well.

But the tokens don't come from another bucket, they come from "nowhere", by time.

BenPope commented on src/v/ssx/include/ssx/sharded_int.h:
```diff
@@ -0,0 +1,69 @@
+
+    /// return the maximum value across the shards
+    ss::future<T> max() const {
+        std::vector<std::atomic<T>> state_copy(ss::smp::count);
```
Why are these atomic?

This function could be called `reduce` and take a reducer.

pgellert replied:
It was to ensure the changes to `state_copy` are visible when we read it, but I think sending the result back through the return value of `submit_to` would have been an alternative. Please correct me if I'm wrong but as far as I understand it, without `atomic<T>` or returning a `future<T>` via `submit_to`, we have no guarantees about the changes inside the lambda being visible to the caller shard.

Anyways, that's more for my curiosity now as I've removed this method and rewrote the `last_time_ms` handling using `map_reduce`.

BenPope commented on src/v/ssx/include/ssx/sharded_int.h:
```diff
@@ -0,0 +1,69 @@
+              this->local(), std::memory_order_seq_cst);
+        });
+
+        T result;
+        for (const auto& x : state_copy) {
+            result = std::max(result, x.load(std::memory_order_seq_cst));
+        }
```
Could use `std::reduce` here.

pgellert replied:
Rewrote `gc()` using `map_reduce` to avoid blocking inside the `sharded_ptr::update`.

BenPope commented on src/v/kafka/server/quota_manager.h:
```diff
@@ -80,16 +81,11 @@ class quota_manager : public ss::peering_sharded_service<quota_manager> {
+        // it is read and written from multiple cores. For delay, we want to
+        // change the behaviour such that we track in connection_ctx whether to
+        // enforce the delay or not.
+        ss::lw_shared_ptr<ssx::sharded_int<uint64_t>> last_seen_ms;
```
I think it would be best to keep this as a `clock::time_point`.

pgellert replied:
Makes sense, it's now a `clock::time_point`.

pgellert commented on src/v/ssx/include/ssx/sharded_int.h:
```diff
@@ -0,0 +1,69 @@
+    T& local() { return _state[ss::this_shard_id()]; }
+
+private:
+    std::vector<T> _state;
```
I think so, I've aligned them now to `ss::cache_line`.

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -350,19 +359,28 @@ void quota_manager::gc() {
+                    auto expire_age = full_window * 10;
+                    for (auto kv : old_map) {
+                        auto last_seen_ms
+                          = co_await kv.second.last_seen_ms->max();
```
Agreed. I rewrote this now to avoid cross-shard calls inside the `sharded_ptr::update` by first gathering a list of expired keys and then updating the map by removing them like you said.

The downside with my current implementation is that there is a small chance that we incorrectly remove a client id that we (1) first think should be GC'd because we haven't seen for a while (2) but then before we get to updating the map another shard touches this client id again and bumps its `last_seen_ms`. Then we would incorrectly remove that client id. But that seems acceptable given that it should be extremely rare (the request from the client id has to come in exactly 10s after its previous request perfectly racing with `gc()`) and even if this happens we will just end up under-throttling the client.

BenPope commented on src/v/kafka/server/quota_manager.h:
```diff
@@ -111,33 +145,16 @@ class quota_manager : public ss::peering_sharded_service<quota_manager> {
       uint32_t mutations,
       clock::time_point now);
 
-    // throttle, return <previous delay, new delay>
-    std::chrono::milliseconds throttle(
+    std::chrono::milliseconds cap_to_max_delay(
```
I think a lot of the `duration_cast` noise can be alleviated by working in `clock::duration`

pgellert replied:
Beautiful 🙌

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -346,27 +348,73 @@ ss::future<throttle_delay> quota_manager::throttle_fetch_tp(
+        return result;
+    };
+
+    auto expired_keys = co_await ss::map_reduce(
+      ss::smp::all_cpus(), mapper, key_set{}, reducer);
```
...leveraging peering_sharded<>::map_reduce
```suggestion
    auto expired_keys = co_await container().map_reduce0(
      mapper, key_set{}, reducer);

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -346,27 +348,73 @@ ss::future<throttle_delay> quota_manager::throttle_fetch_tp(
+    };
+
+    auto reducer = [](const key_set& acc, const key_set& next) -> key_set {
+        key_set result;
+        // Note: std::set_intersection assumes the inputs are sorted and
+        // guarantees that the output is sorted
+        std::set_intersection(
+          acc.begin(),
+          acc.end(),
+          next.begin(),
+          next.end(),
+          std::back_inserter(result));
+        return result;
```
It might be worth benchmarking this, there are potentially lots of allocs.

It might be cheaper to take the parameters by value (I think they're moved in), and remove from `acc` the values in `next`.

oleiman replied:
Any word on this? Assuming there are copies made at the `reducer` call site (in `mp_reduce0`), I don't see why we wouldn't compute the intersection in place.

pgellert replied:
Yeah I haven't yet benchmarked gc specifically, but passing by value and doing the intersection in-place makes sense. Let me take this up as a follow up performance improvement PR though to unblock Ben's work on dynamic quota changes.

oleiman replied:
sgtm

pgellert commented on src/v/kafka/server/connection_context.cc:
```diff
@@ -795,74 +795,75 @@ ss::future<> connection_context::client_protocol_state::handle_response(
+    connection_ctx->_server.handler_probe(request_key)
+      .add_bytes_sent(response_size);
+    try {
+        co_return co_await connection_ctx->conn->write(std::move(msg))
+          .then([] {
+              return ss::make_ready_future<ss::stop_iteration>(
+                ss::stop_iteration::no);
+          })
+          // release the resources only once it has been written to
+          // the connection.
+          .finally([resources = resp_and_res.resources] {});
```
Nice catch

pgellert commented on src/v/kafka/server/atomic_token_bucket.h:
```diff
@@ -0,0 +1,92 @@
+    using bucket_t = ss::internal::shared_token_bucket<
+      uint64_t,
+      std::ratio<1>,
+      ss::internal::capped_release::yes,
```
Yeah it should be the template parameter. Fixed

pgellert commented on src/v/kafka/server/connection_context.cc:
```diff
@@ -529,8 +533,8 @@ ss::future<session_resources> connection_context::throttle_request(
     // affect.
-
-    const delay_t delay = record_tp_and_calculate_throttle(hdr, request_size);
+    const delay_t delay = co_await record_tp_and_calculate_throttle(
```
Nice catch. I've changed this so that we copy around `request_data` for quotas instead of taking a reference to the header.

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -36,55 +58,104 @@ quota_manager::quota_manager()
-             static_cast<uint32_t>(_default_num_windows()),
-             _default_window_width()})
-          : std::optional<token_bucket_rate_tracker>()});
+    auto qid = ss::sstring{quota_id ? *quota_id : ""};
```
Nice, yeah this worked

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -36,55 +58,104 @@ quota_manager::quota_manager()
+        auto replenish_threshold = static_cast<uint64_t>(
+          _replenish_threshold().value_or(1));
+
+        auto update_func =
+          [produce_rate,
+           fetch_rate,
+           partition_mutation_rate,
+           default_num_windows,
+           default_window_width,
+           replenish_threshold,
+           qid,
+           now](const client_quotas_map_t& old_map) -> client_quotas_map_t {
```
Yep, that's nicer

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -36,55 +58,104 @@ quota_manager::quota_manager()
+    auto it = _client_quotas->find(qid);
+    auto inserted = false;
+    if (it == _client_quotas->end()) {
+        auto produce_rate = get_client_target_produce_tp_rate(qid);
+        auto fetch_rate = get_client_target_fetch_tp_rate(qid);
+        auto partition_mutation_rate = *_target_partition_mutation_quota();
+        auto default_num_windows = static_cast<uint32_t>(
+          _default_num_windows());
+        auto default_window_width = _default_window_width();
+        auto replenish_threshold = static_cast<uint64_t>(
+          _replenish_threshold().value_or(1));
```
I was trying to avoid capturing `this`. But we're already taking a reference to the sharded map as a parameter, so we're already relying on `quota_manager` to outlive this section of the code, so I went ahead and moved these into the `update_func` to simplify.

It's worth noting that moving this into the `update_func` means that we now spend a few more cycles inside the mutex of the `sharded_ptr`. But I expect the map copying to dominate the time inside the `update_func` so this should be negligible.

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -36,55 +58,104 @@ quota_manager::quota_manager()
+            // the handler core, gc() gets scheduled and cleans up the inserted
+            // quota, and by the time the handler gets here the inserted entry
+            // is gone. In that case, we give up.
+            vlog(klog.debug, "Failed to find quota id after insert...");
+            throw std::runtime_error("Failed to find quota id after insert");
```
Based on some local testing I've done, we close the connection in this case and the client has to reconnect.
```
DEBUG 2024-05-17 09:45:38,776 [shard 0:main] kafka - server.cc:310 - max_reauth_ms: 0 // <-- Kafka accepts the connection
DEBUG 2024-05-17 09:45:38,777 [shard 0:main] kafka - connection_context.cc:125 - Connection input_shutdown; aborting operations for 172.26.0.29:38986
DEBUG 2024-05-17 09:45:38,778 [shard 0:main] kafka - connection_context.cc:125 - Connection input_shutdown; aborting operations for 172.26.0.29:38992
WARN  2024-05-17 09:45:38,779 [shard 0:main] kafka - server.cc:407 - Error kafka rpc protocol 172.26.0.29:38992: std::runtime_error (Failed to find quota id after insert)
ERROR 2024-05-17 09:45:38,779 [shard 0:main] kafka - server.cc:159 - Error[applying protocol] remote address: 172.26.0.29:38992 - std::runtime_error (Failed to find quota id after insert)
DEBUG 2024-05-17 09:45:38,883 [shard 0:main] kafka - server.cc:310 - max_reauth_ms: 0 // <-- Kafka accepts the re-connection
```
I think that's fine because we really don't expect this to happen. The post-await code here needs to be delayed for the delay interval (~10s by default) and needs to be reordered with respect to the update done by `gc()`. So I'm not even sure it's possible because I would assume that the scheduler would not reorder this code wrt updates issued later.

But the alternatives I can think of are (a) retrying here more to reduce the chances of this happening further, or (b) handling this failure one stack up but still within `quota_manager` and just not applying any delays in this case. Both of these make the code more complex for the sake of this rare case, but I guess (b) would be the most reliable way of handling this (limiting failure to quotas). What do you think?

pgellert replied:
I've changed this now so that we just don't throttle in this case.

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -346,27 +348,73 @@ ss::future<throttle_delay> quota_manager::throttle_fetch_tp(
+          for (auto& k : expired_keys) {
+              new_map.erase(k);
+          }
+          co_return new_map;
```
I didn't have a copy of the map, only a const ref to it. But I've fixed that now by doing the copy inside `sharded_ptr::update` and after that applied your suggestions.

pgellert commented:
And [another small one](https://github.com/redpanda-data/redpanda/compare/4abc93be3baac29af3434dd451e1b3cd74c3df91..e727cf97f57ac2c6278c1680efc4c9095e4f17a2) to fix the linting error.

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -36,54 +61,56 @@ quota_manager::quota_manager()
+    auto it = _client_quotas->find(qid);
+    auto inserted = false;
+    if (it == _client_quotas->end()) {
+        co_await container().invoke_on(
+          _client_quotas.shard_id(),
+          [qid, now](quota_manager& me) { return me.add_quota_id(qid, now); });
```
I think you also need to hold a gate on both sides (the calling shard and the called shard), so perhaps it's worth moving the dispatch and gates holding into `add_quota_id` and avoiding the assert.

pgellert replied:
The `_gate` is only used to fail the `gc()` fibre once `quota_manager.stop()` has been called, so I don't think we need to hold gates on the request path. On the request path, we can rely on the assumption that `quota_manager` is only stopped/destructed after any outstanding requests have been processed/failed:
https://github.com/redpanda-data/redpanda/blob/a7cd2b0de8fd494bf0c49292b658900416efff22/src/v/redpanda/application.cc#L321-L338

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -93,6 +120,47 @@ quota_manager::maybe_add_and_retrieve_quota(
+            produce_tracker,
+            fetch_tracker,
+            /// pm_rate is only accessed on the qm home shard
+            std::optional<token_bucket_rate_tracker>(
+              {partition_mutation_rate,
+               default_num_windows,
+               default_window_width})});
```
```suggestion
            std::make_optional<token_bucket_rate_tracker>(
              partition_mutation_rate,
              default_num_windows,
              default_window_width)});
```

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -36,55 +55,114 @@ quota_manager::quota_manager()
-        auto full_window = _default_num_windows() * _default_window_width();
-        gc(full_window);
-    });
+    _gc_timer.set_callback([this]() { gc(); });
```
I don't think it matters semantically, but I'll remove it to make the code clearer

pgellert commented:
[Rebased to dev](https://github.com/redpanda-data/redpanda/compare/e727cf97f57ac2c6278c1680efc4c9095e4f17a2..6769a080a6510448a090be43c646b6a7565eb736). I'll address the code review above next.

pgellert commented on src/v/kafka/server/connection_context.cc:
```diff
@@ -821,8 +821,7 @@ connection_context::client_protocol_state::do_process_responses(
         co_await connection_ctx->_server.quota_mgr().record_fetch_tp(
-          resp_and_res.resources->request_data.client_id.value_or(""),
-          msg.size());
+          resp_and_res.resources->request_data.client_id, msg.size());
```
The previous commit tried to be just a refactor, so because it changed the type of `request_data.client_id` from `ss::sstring` to `std::optional<ss::sstring>`, to maintain the behaviour I had to do the `value_or` here. The next commit then removes this `value_or`. I'll fold this in though, I mainly just wanted to make this change apparent.

pgellert commented on src/v/ssx/include/ssx/sharded_ptr.h:
```diff
@@ -84,6 +85,31 @@ class sharded_ptr {
+        }
+
+        auto u = std::make_shared<T>(co_await ss::futurize_invoke(
+          std::forward<F>(update_func), local() ? T{*local()} : T{}));
```
Yeah, let me unroll this

pgellert commented on src/v/kafka/server/connection_context.cc:
```diff
@@ -514,35 +515,32 @@ connection_context::record_tp_and_calculate_throttle(
-    return fut
-      .then([this, key = hdr.key, request_size] {
+    auto track = track_latency(r_data.request_key);
+    co_return co_await fut
```
Makes sense, I've added a new commit for this refactor

pgellert commented on src/v/kafka/server/quota_manager.h:
```diff
@@ -66,7 +72,36 @@ class quota_manager : public ss::peering_sharded_service<quota_manager> {
+        std::optional<token_bucket_rate_tracker> pm_rate;
+    };
+
+    using client_quotas_map_t = absl::
+      node_hash_map<ss::sstring, client_quota, sstring_hash, sstring_eq>;
```
Thanks, yeah this is what we need.

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -250,36 +250,36 @@ throttle_delay quota_manager::record_produce_tp_and_throttle(
     }
-    auto it = maybe_add_and_retrieve_quota(quota_id, now);
+    auto it = co_await maybe_add_and_retrieve_quota(quota_id, now);
     it->second.tp_fetch_rate.maybe_advance_current(now);
```
Used your suggestion from here: https://github.com/redpanda-data/redpanda/pull/18098#discussion_r1605041327

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -250,36 +250,36 @@ throttle_delay quota_manager::record_produce_tp_and_throttle(
       client_id, _target_fetch_tp_rate_per_client_group());
-    auto it = maybe_add_and_retrieve_quota(quota_id, now);
+    auto it = co_await maybe_add_and_retrieve_quota(quota_id, now);
     it->second.tp_fetch_rate.record(bytes, now);
```
Used your suggestion from here: https://github.com/redpanda-data/redpanda/pull/18098#discussion_r1605041327

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -233,13 +233,13 @@ std::chrono::milliseconds quota_manager::throttle(
-    auto it = maybe_add_and_retrieve_quota(quota_id, now);
+    auto it = co_await maybe_add_and_retrieve_quota(quota_id, now);
 
     it->second.tp_produce_rate.record(bytes, now);
```
Used your suggestion from here: https://github.com/redpanda-data/redpanda/pull/18098#discussion_r1605041327

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/6769a080a6510448a090be43c646b6a7565eb736..ad0fc7434411b366ab17552e7d21109befd5af96) with more code review changes

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/ad0fc7434411b366ab17552e7d21109befd5af96..10d3124d036be41ac25ff7b79b95b8882c508415) with a microbenchmark in the last commit and some bug fixes.

#### Benchmark
I've now added a microbenchmark to the fetch-related methods of quota manager to try to measure the worst-case performance of these methods under high cross-core contention. The worst case seems to be when all the client ids are the same and so all the cores are contending on the atomic accesses to the client quotas. Even in this case we could handle ~10M requests in ~1.427s which is about 136ns per request. This is certainly more than what shard-local access would be, but it is still a relatively low number that should be within our performance budget for broker-wide client quota tracking.

```
single run iterations:    0
single run duration:      1.000s
number of runs:           5
number of cores:          32
random seed:              4113225649

test                                                iterations      median         mad         min         max      allocs       tasks        inst
quota_group.test_quota_manager_on_unlimited_shared           1      1.427s     8.987ms      1.410s      1.436s 1638484.800    2360.200         0.0
quota_group.test_quota_manager_on_unlimited_unique           6   221.734ms     9.034ms   209.158ms   230.804ms 2034272.233    9752.833         0.0
quota_group.test_quota_manager_on_limited_shared             1      1.156s     1.515ms      1.153s      1.158s 1638485.000    1590.600         0.0
quota_group.test_quota_manager_on_limited_unique             6   187.928ms     1.592ms   186.336ms   229.697ms 2033969.700    9680.200         0.0
quota_group.test_quota_manager_off_shared                    1      1.254s     2.329ms      1.252s      1.268s 1638484.200    2773.200         0.0
quota_group.test_quota_manager_off_unique                    6   165.436ms     5.238ms   160.198ms   214.782ms 2034187.600    9702.833         0.0
```

#### Bug fix
Furthermore, in this force push, I also fixed a bug to do with the lifetime of the `client_quota` objects. We need to ensure that the `client_quota` objects returned from the map are safe to hold on to across scheduling points and that even if the `client_quota` is removed from the map during `gc()`, the last desctruction of the `client_quota` object should happen on shard 0. For this, I've introduced a reference wrapped (`value_node`) to ensure exactly this using `std::shared_ptr` and `ss::foreign_ptr`.

BenPope commented on src/v/kafka/server/quota_manager.h:
```diff
@@ -58,7 +65,60 @@ class quota_manager : public ss::peering_sharded_service<quota_manager> {
+    struct client_quota {
+        ssx::sharded_value<clock::time_point> last_seen_ms;
+        // TODO: do we need to pad these?
+        std::optional<bucket_t> tp_produce_rate;
```
`rate_tracker can be deleted, I think.

pgellert replied:
Yes, I've removed it now.

BenPope commented on src/v/kafka/server/quota_manager.h:
```diff
@@ -58,7 +65,60 @@ class quota_manager : public ss::peering_sharded_service<quota_manager> {
+        // TODO: do we need to pad these?
+        std::optional<bucket_t> tp_produce_rate;
+        std::optional<bucket_t> tp_fetch_rate;
+        std::optional<token_bucket_rate_tracker> pm_rate;
```
I think it's worth doing this one, too, and removing the dispatch to home core.  And then `token_bucket_rate_tracker` can be removed.

pgellert replied:
I've made this change now

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -162,8 +231,11 @@ std::chrono::milliseconds quota_manager::do_record_partition_mutations(
     auto quota_id = get_client_quota_id(client_id, {});
-    auto it = maybe_add_and_retrieve_quota(quota_id, now);
-    const auto units = it->second.pm_rate->record_and_measure(mutations, now);
+    auto cq = co_await maybe_add_and_retrieve_quota(quota_id, now);
+    if (!cq) {
+        co_return std::chrono::milliseconds::zero();
+    }
+    const auto units = cq.value()->pm_rate->record_and_measure(mutations, now);
```
As discussed, we can probably void the copy of what was previously a seastar shared_pointer by moving to an interface that encapsulates this:
```c++
    return record_and_measure_pm(quota_id, mutations, now);
```

pgellert replied:
I've pushed more changes now for this as well: https://github.com/redpanda-data/redpanda/pull/18098#issuecomment-2124940958.

The interface for this is really the interface of the `quota_manager`, eg. `record_partition_mutations(client_id, mutations, now)`. I haven't found a better interface to wrap the map in but I am open to suggestions.

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/10d3124d036be41ac25ff7b79b95b8882c508415..25227e5829985a2e05ad9c16a5ef5642a1a5ffa6) further changes based on code review and local testing:
* The partition mutation quotas now also use the atomic token buckets. This also required some changes to the tests (see the commit message).
* I've changed all uses of the `atomic_token_bucket` to be with `capped_release::no`, because we don't need it. After some testing I realized that we need `capped_release::no` since we don't need to hold onto units from the token bucket and restrict the releasing of tokens back to the token bucket. I was prompted to do this by the partition mutation ducktape tests as I noticed that the token bucket wasn't getting replenished because we never called `release()` on the token capped token bucket. I've now added more tests around both the `atomic_token_bucket` and `quota_manager` to exercise more of the behaviour.
* I've added a check into the throttling methods to only insert into the quotas map if throttling is configured for the client. This helps with the case when quotas are turned off, as shown by the updated benchmark (when quotas are off, quota management has a ~2.5ns overhead.
* Deleted a bunch of unused code 🔥

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/25227e5829985a2e05ad9c16a5ef5642a1a5ffa6..d217383b5285c8c0c868bd0d2f010d5b6e525192) with changes to avoid the `sharded_ptr` copy on the request path and avoid the `value_node` in the map. Improved benchmark:
```
test                                                iterations      median         mad         min         max      allocs       tasks        inst
quota_group.test_quota_manager_on_unlimited_shared           3   457.729ms   171.802us   457.558ms   458.047ms 2293843.600     685.067         0.0
quota_group.test_quota_manager_on_unlimited_unique           5   192.070ms     3.564ms   186.680ms   198.077ms 2032395.040    9758.880         0.0
quota_group.test_quota_manager_on_limited_shared             3   474.613ms   305.042us   473.773ms   474.918ms 2293843.200     702.533         0.0
quota_group.test_quota_manager_on_limited_unique             6   197.639ms     2.688ms   193.524ms   200.552ms 2032468.133    9730.600         0.0
quota_group.test_quota_manager_off_shared                   41    24.975ms   195.504us    24.780ms    28.671ms 1638437.693      58.068         0.0
quota_group.test_quota_manager_off_unique                   49    20.691ms   254.579us    20.418ms    21.071ms  655397.441      47.833         0.0
```

BenPope commented on src/v/ssx/include/ssx/sharded_ptr.h:
```diff
@@ -84,6 +85,34 @@ class sharded_ptr {
+               ss::future<T>,
+               ss::futurize_t<std::invoke_result_t<F, T>>>
+    ss::future<> update(F&& update_func) {
+        assert_shard();
+        auto mu{co_await _mutex.get_units()};
+        if (_state.empty()) {
+            _state.resize(ss::smp::count);
+        }
+
+        auto new_value = local() ? T{*local()} : T{};
+        auto updated_new_value = co_await ss::futurize_invoke(
+          std::forward<F>(update_func), std::move(new_value));
+        auto new_ref = std::make_shared<T>(std::move(updated_new_value));
+        auto copy_to_deallocate_on_owner_shard = local();
+        co_await ss::smp::invoke_on_all(
+          [this, new_ref]() noexcept { local() = new_ref; });
```
I don't think this needs to be async, but a simplification would seem to be:
```suggestion
        co_return co_await reset(co_await ss::futurize_invoke(
          std::forward<F>(update_func),
          local() ? std::as_const(*local()) : T{}));
```

BenPope reviewed (commented):
I have some cleanups here and tested out inverting the control so that we don't have to worry about the lifetime of the iterator / shared_object to client_quote: https://github.com/BenPope/redpanda/commit/ca42011f9d889504f2a5c59739dbda521b14a830

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/d217383b5285c8c0c868bd0d2f010d5b6e525192..954c5f981d11b076987a8f7c83f6af8ab668cafb) to adopt the suggestions above.

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -147,105 +201,66 @@ ss::future<std::chrono::milliseconds> quota_manager::record_partition_mutations(
+    std::chrono::milliseconds delay_ms
+      = std::chrono::duration_cast<std::chrono::milliseconds>(delay);
     if (delay_ms > max_delay_ms) {
         vlog(
           klog.info,
-          "Found data rate for window of: {} bytes. Client:{}, "
+          "Client:{}, "
           "Estimated "
           "backpressure delay of {}. Limiting to {} backpressure delay",
-          rate,
           quota_id,
           delay_ms,
           max_delay_ms);
-        delay_ms = max_delay_ms;
+        delay = max_delay_ms;
     }
-    return delay_ms;
+    return delay;
```
I know ultimately that it'll end up as milliseconds, but let's keep precision, just in case.
```suggestion
    clock::duration max_delay = _max_delay();
    if (delay > max_delay) {
        vlog(
          klog.info,
          "Client:{}, "
          "Estimated "
          "backpressure delay of {}. Limiting to {} backpressure delay",
          quota_id,
          std::chrono::duration_cast<std::chrono::milliseconds>(delay),
          std::chrono::duration_cast<std::chrono::milliseconds>(max_delay));
    }
    return std::min(delay, max_delay);
```

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -254,37 +269,67 @@ ss::future<> quota_manager::record_fetch_tp(
+    vassert(
+      ss::this_shard_id() == _client_quotas.shard_id(),
+      "gc should only be performed on the owner shard");
+    auto full_window = _default_num_windows() * _default_window_width();
```
Ahh, this is a shame, otherwise the config options are unused.

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -254,37 +269,67 @@ ss::future<> quota_manager::record_fetch_tp(
+          [full_window, _client_quotas = std::ref(_client_quotas)]() {
+              auto now = clock::now();
+              auto expire_age = full_window * 10;
+              auto new_map = client_quotas_map_t{*_client_quotas.get().local()};
```
This copy is unfortunate, but I see it goes away in a future commit.

pgellert replied:
I've reworked that commit now to just comment out the gc code and have gc implemented only once in the later commit.

BenPope replied:
> I've reworked that commit now to just comment out the gc code and have gc implemented only once in the later commit.

That's probably ok, but I guess some more reordering of commits could avoid it.

BenPope reviewed (commented):
Looking pretty nice now!

pgellert commented on src/v/ssx/include/ssx/sharded_ptr.h:
```diff
@@ -84,6 +85,20 @@ class sharded_ptr {
+    template<typename F>
+    requires std::invocable<F, T> && std::same_as<T, std::invoke_result_t<F, T>>
+    ss::future<> update(F&& update_func) {
+        co_return co_await reset(
+          update_func(local() ? std::as_const(*local()) : T{}));
```
Updated. I've also now realised that I had to switch up update and reset such that reset is implemented in terms of the update and not the other way around, because with `reset(update_func(local() ? std::as_const(*local()) : T{}))` the update happens outside of the mutex.

BenPope replied:
> Updated. I've also now realised that I had to switch up update and reset such that reset is implemented in terms of the update and not the other way around, because with `reset(update_func(local() ? std::as_const(*local()) : T{}))` the update happens outside of the mutex.

The `update_func` is called outside of the mutex, but that's ok, since it's just an observer, it's either taking a newly constructed value or a const&, so no changes can be made to the underlying data structure.

The actual shard updates are done with the returned value passed to reset, which takes the mutex.

pgellert commented on src/v/kafka/server/quota_manager.h:
```diff
@@ -54,11 +56,27 @@ class quota_manager : public ss::peering_sharded_service<quota_manager> {
+    // pm_rate: partition mutation quota tracking
+    struct client_quota {
+        clock::time_point last_seen;
+        // TODO: do we need to pad these?
```
Let's not pad for now then

BenPope replied:
> Let's not pad for now then

The benchmark should be useful to determine whether aligning them is beneficial.

pgellert replied:
I've added padding now based on the modified benchmark (I've pushed the benchmark change as well): https://github.com/redpanda-data/redpanda/pull/18098#issuecomment-2135080127

pgellert commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -37,60 +49,102 @@ quota_manager::quota_manager()
+        auto new_value = ss::make_lw_shared<client_quota>(
+          now, std::nullopt, std::nullopt, std::nullopt);
+
+        new_value->tp_produce_rate.emplace(
```
The produce throttling config `target_quota_byte_rate` is not an `optional<int>` but an `int` so it is always on. It looks like it was initially added like that (https://github.com/redpanda-data/redpanda/commit/4e0eeb31435189b59e089b8d62304dd18800c0d1) and then we kept increasing the limit default (eg. https://github.com/redpanda-data/redpanda/commit/2e47c7a5d83d99d9e60d72f9776b8a7f6b8b92a0, https://github.com/redpanda-data/redpanda/commit/1d8200a9f8f1e980a50ebe9dbb93ac7d26cc3c8e).

I wonder if we should change `target_quota_byte_rate` to drop the minimum to 0, interpret 0 as disabled and make 0 the default. What do you think?

BenPope replied:
Does it not being optional imply that we always track for all clients?

pgellert replied:
Yes

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -36,61 +50,105 @@ quota_manager::quota_manager()
+
+    auto update_func = [this, qid = ss::sstring{qid}, now](
+                         client_quotas_map_t new_map) -> client_quotas_map_t {
+        auto produce_rate = get_client_target_produce_tp_rate(qid);
+        auto fetch_rate = get_client_target_fetch_tp_rate(qid);
+        auto partition_mutation_rate = _target_partition_mutation_quota();
```
Do these all need multiplying by ss::smp::count?

pgellert replied:
Yes, but let's do that as a separate PR because the ducktape tests will need to be adjusted for that. I'll need to think about how we can make the ducktape tests work. We might need to set the configs based on the core count of the nodes (if that's something we know from the ducktape side) or maybe move the tests to use the new quota store once that's available.

Created https://redpandadata.atlassian.net/browse/CORE-3078 to track this.

BenPope replied:
> Yes, but let's do that as a separate PR because the ducktape tests will need to be adjusted for that.

I'm surprised that changing the quota by a factor of `ss::smp::count` doesn't already break the tests,

pgellert replied:
It will break when we haven't done the upscaling yet. I think all the connections in the tests end up on the same node+shard, so the tests today work quite similarly to how they did before (they enforce the same limits, just using atomic instructions).

BenPope replied:
Ooof.

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/954c5f981d11b076987a8f7c83f6af8ab668cafb..f7dbb46559f4ff14c732f1bf4773985fe1f92143) with more suggested changes. I have yet to benchmark and address the point about adding padding around the token buckets.

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/49526#018fab34-2027-4ad1-a6ed-9e84e8cfe928:
```
"rptest.tests.usage_test.UsageTestCloudStorageMetrics.test_usage_manager_cloud_storage"
```
<!--8d5d19a8-ffea-4fba-add2-4569e2642cd6-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/49526#018fab34-2027-4ad1-a6ed-9e84e8cfe928
<!--fe317f4d-a847-4c5f-8640-2989026702a6-->

michael-redpanda reviewed (dismissed):
Thanks for going over this in our meeting.  Just a question but in general this looks great.

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -53,44 +65,86 @@ ss::future<> quota_manager::stop() {
+      "add_quota_id should only be called on the owner shard");
+
+    auto update_func = [this, qid = ss::sstring{qid}, now](
+                         client_quotas_map_t new_map) -> client_quotas_map_t {
```
> Oh I guess that's the signature expected. Anyways leaving this here for posterity

Yeah. Either a T&& or T const&, taking a sink param/copy  at call-site to modify in-plaice.

pgellert replied:
It took a copy of the map because in a previous iteration of this PR `update` took an async function. I can revert to `T const&` though now that `update` just takes a normal (non-async) function to simplify.

BenPope commented on src/v/ssx/include/ssx/sharded_ptr.h:
```diff
@@ -84,6 +81,40 @@ class sharded_ptr {
+    ss::future<> update(F&& update_func) {
+        return update_shared([update_func{std::forward<F>(update_func)}](
+                               const std::shared_ptr<T>& v) {
+            return std::make_shared<T>(update_func(v ? *v : T{}));
```
The `update_func` should not be allowed to take a non-const reference to the state.
```suggestion
            return std::make_shared<T>(
              update_func(v ? std::as_const(*v) : T{}));
```

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/f7dbb46559f4ff14c732f1bf4773985fe1f92143..7371aaf23d0bcf962b8b8f8e1df226f523c9477a) with:
* Addressing the comment about passing the argument by `const&` to the close of the `sharded_ptr::update` function
* Added padding around the token buckets to align them to the cache line as the benchmarks confirmed that this leads to a performance improvement when multiple cores contend on neighbouring token buckets. This leads to ~25% improvement in the pathological case of having 16-16 cores updating the neighbouring fetch and produce token buckets.

```
Without cache alignment
test                                                iterations      median         mad         min         max      allocs       tasks        inst
quota_group.test_quota_manager_on_unlimited_shared           3   402.589ms     4.656ms   361.566ms   407.246ms 2293843.800    1231.467         0.0
quota_group.test_quota_manager_on_unlimited_unique           7   165.404ms     1.589ms   162.938ms   169.582ms 2033037.571    9595.000         0.0
quota_group.test_quota_manager_on_limited_shared             3   343.729ms   805.020us   342.924ms   346.064ms 2293843.467    1065.067         0.0
quota_group.test_quota_manager_on_limited_unique             7   164.961ms   276.672us   164.685ms   166.071ms 2033145.971    9587.514         0.0
quota_group.test_quota_manager_off_shared                    5   223.623ms     1.085ms   222.348ms   225.867ms 1639174.400     208.880         0.0
quota_group.test_quota_manager_off_unique                    9   111.680ms   719.302us   110.740ms   112.762ms 1448054.867    9181.511         0.0

With cache alignment
test                                                iterations      median         mad         min         max      allocs       tasks        inst
quota_group.test_quota_manager_on_unlimited_shared           4   302.860ms   869.473us   301.991ms   306.604ms 2293843.950    1022.550         0.0
quota_group.test_quota_manager_on_unlimited_unique           7   161.132ms   975.600us   160.103ms   164.485ms 2032651.086    9590.971         0.0
quota_group.test_quota_manager_on_limited_shared             4   307.775ms   356.912us   304.962ms   308.553ms 2293843.500    1089.250         0.0
quota_group.test_quota_manager_on_limited_unique             7   160.303ms     1.421ms   155.586ms   169.324ms 2032743.971    9594.314         0.0
quota_group.test_quota_manager_off_shared                    5   248.071ms     3.099ms   231.620ms   256.416ms 1639174.360     207.240         0.0
quota_group.test_quota_manager_off_unique                    6   162.579ms    10.868ms   149.720ms   256.007ms 1448036.300    9227.067         0.0
```

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -310,17 +314,73 @@ void quota_manager::gc() {
+    // between the two steps can cause recently seen client ids to be
+    // deleted. That's acceptable because this should be rare and the client
+    // will be tracked correctly again from when we next see it.
+    co_await _client_quotas.update([expired_keys{std::move(expired_keys)}](
+                                     const client_quotas_map_t& old_map) {
+        auto new_map = client_quotas_map_t{old_map};
```
Why not take the map by value to get the copy?
```suggestion
    co_await _client_quotas.update(
      [expired_keys{std::move(expired_keys)}](client_quotas_map_t new_map) {

pgellert replied:
This is now passed by const ref (see the last force push).

pgellert replied:
I see what you mean now after our chat offline. Fixed this now: https://github.com/redpanda-data/redpanda/pull/18098#issuecomment-2135493570

BenPope commented on src/v/kafka/server/quota_manager.cc:
```diff
@@ -53,44 +65,88 @@ ss::future<> quota_manager::stop() {
+      ss::this_shard_id() == _client_quotas.shard_id(),
+      "add_quota_id should only be called on the owner shard");
+
+    auto update_func =
+      [this, qid = ss::sstring{qid}, now](
+        const client_quotas_map_t& old_map) -> client_quotas_map_t {
```
Why not take the map by value to get the copy?
```suggestion
    auto update_func =
      [this, qid = ss::sstring{qid}, now](client_quotas_map_t new_map) {

pgellert replied:
This is now passed by const ref (see the last force push).

pgellert replied:
Fixed: https://github.com/redpanda-data/redpanda/pull/18098#issuecomment-2135493570

oleiman commented on src/v/kafka/server/quota_manager.h:
```diff
@@ -54,11 +56,26 @@ class quota_manager : public ss::peering_sharded_service<quota_manager> {
     };
 
-    quota_manager();
+    using client_quotas_map_t = absl::node_hash_map<
+      ss::sstring,
+      ss::lw_shared_ptr<client_quota>,
+      sstring_hash,
+      sstring_eq>;
+    using client_quotas_t = ssx::sharded_ptr<client_quotas_map_t>;
+
+    explicit quota_manager(client_quotas_t& client_quotas);
```
Just want to restate the argument for why the mutable `lw_shared` is safe here: off-owner mutation is always done through two layers of indirection - an iterator into the underlying map, which holds a _reference_ to the `lw_shared_ptr`, which is then deref'ed into a mutable `client_quota&` and passed to the mutation callback. Do I have that right?

Q: when will the ref count on the `lw_shared_ptr<client_quota>` (the quota map value itself) be `>1`? Only when you hold a copy of the map itself in an update func for `sharded_ptr`?

pgellert replied:
Yes, that's all correct.

oleiman replied:
nice 👍

oleiman reviewed (approved):
I've read and (I think) understood this and LGTM. Will probably take another pass later today or tomorrow with the benefit of hindsight.
